core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -4769,21 +4779,47 @@ private void convertValuesImpl(
       return;
     }
 
-    for (SqlNode rowConstructor1 : values.getOperandList()) {
-      SqlCall rowConstructor = (SqlCall) rowConstructor1;
+    SqlCall insertOp = callStack.peek();
+    boolean processDefaults = insertOp != null && DefaultChecker.check(values);
+
+    if (targetRowType == null) {
+      RelDataType listType = validator().getValidatedNodeType(values);
+      targetRowType = SqlTypeUtil.promoteToRowType(typeFactory, listType, null);
+    }
+
+    assert insertOp instanceof SqlInsert || !processDefaults : "operation: " + insertOp;
+
+    @Nullable InitializerExpressionFactory initializerFactory = processDefaults
+            ? getInitializerFactory(getNamespace(requireNonNull(insertOp, "insertOp")).getTable())
+        : null;
+    @Nullable RelOptTable targetTable = processDefaults
+        ? getTargetTable(requireNonNull(insertOp, "insertOp")) : null;
+    int[] mapping = processDefaults
+        ? processColumnsMapping(requireNonNull(targetRowType, "targetRowType"),
+        requireNonNull(targetTable, "targetTable")) : null;
```

Comment:
This code block is hard to read. All 3 variables have `processDefaults ?` in their expression, and your inconsistent formatting doesn't help. One `if (processDefaults)` block would be easier to read.